# Task

I made this notebook to investigate issues with high model perplexity while reducting loss

I use this model: https://huggingface.co/TheBloke/vicuna-13B-1.1-GPTQ-4bit-128g/tree/main

## Implementation

In [1]:
import torch
import numpy as np

In [2]:
from alpaca_lora_4bit.monkeypatch.peft_tuners_lora_monkey_patch import replace_peft_model_with_int4_lora_model
replace_peft_model_with_int4_lora_model()

c:\Users\alex4321\anaconda3\envs\llama\Lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


Triton not found. Please run "pip install triton".


In [3]:
from alpaca_lora_4bit import autograd_4bit
autograd_4bit.switch_backend_to('cuda')

Using CUDA implementation.


In [4]:
from alpaca_lora_4bit.autograd_4bit import load_llama_model_4bit_low_ram
from peft import LoraConfig, get_peft_model, get_peft_model_state_dict, PeftModel, set_peft_model_state_dict

In [5]:
model, tokenizer = load_llama_model_4bit_low_ram("vicuna-13b-GPTQ-4bit-128g",
                                                 "vicuna-13b-GPTQ-4bit-128g/vicuna-13b-4bit-128g.safetensors",
                                                 groupsize=128,
                                                 is_v1_model=False)

Loading Model ...


The safetensors archive passed at vicuna-13b-GPTQ-4bit-128g/vicuna-13b-4bit-128g.safetensors does not contain metadata. Make sure to save your model with the `save_pretrained` method. Defaulting to 'pt' metadata.


Loaded the model in 6.80 seconds.


In [6]:
TEXTS = ["""
The vicuña (Lama vicugna) or vicuna[3] (both /vɪˈkuːnjə/, very rarely spelled vicugna, its former genus name) is one of the two wild South American camelids, which live in the high alpine areas of the Andes, the other being the guanaco, which lives at lower elevations.
Vicuñas are relatives of the llama, and are now believed to be the wild ancestor of domesticated alpacas, which are raised for their coats.
Vicuñas produce small amounts of extremely fine wool, which is very expensive because the animal can only be shorn every three years and has to be caught from the wild.
When knitted together, the product of the vicuña's wool is very soft and warm.
The Inca valued vicuñas highly for their wool, and it was against the law for anyone but royalty to wear vicuña garments;
today, the vicuña is the national animal of Peru and appears on the Peruvian coat of arms.

Both under the rule of the Inca and today, vicuñas have been protected by law, but they were heavily hunted in the intervening period.
At the time they were declared endangered in 1974, only about 6,000 animals were left.
Today, the vicuña population has recovered to about 350,000, and although conservation organizations have reduced its level of threat classification, they still call for active conservation programs to protect populations from poaching, habitat loss, and other threats.

Previously the vicuña was thought not to have been domesticated, and the llama and the alpaca were both regarded as descendants of the closely related guanaco.
But DNA research published in 2001 has shown the alpaca may well have vicuña parentage.
Today, the vicuña is mainly wild, but the local people still perform special rituals with these creatures, including a fertility rite.
""".strip(),
"""
The llama (/ˈlɑːmə/; Spanish pronunciation: [ˈʎama]) (Lama glama) is a domesticated South American camelid, widely used as a meat and pack animal by Andean cultures since the Pre-Columbian era.

Llamas are social animals and live with others as a herd.
Their wool is soft and contains only a small amount of lanolin.
Llamas can learn simple tasks after a few repetitions.
When using a pack, they can carry about 25 to 30% of their body weight for 8 to 13 km (5–8 miles).
The name llama (in the past also spelled "lama" or "glama") was adopted by European settlers from native Peruvians.

The ancestors of llamas are thought to have originated from the Great Plains of North America about 40 million years ago, and subsequently migrated to South America about three million years ago during the Great American Interchange.
By the end of the last ice age (10,000–12,000 years ago), camelids were extinct in North America.
As of 2007, there were over seven million llamas and alpacas in South America and over 158,000 llamas and 100,000 alpacas, descended from progenitors imported late in the 20th century, in the United States and Canada.[5]

In Aymara mythology, llamas are important beings.
The Heavenly Llama is said to drink water from the ocean and urinates as it rains.
According to Aymara eschatology, llamas will return to the water springs and lagoons where they come from at the end of time.
""".strip(),
"""
The alpaca (Lama pacos) is a species of South American camelid mammal.
It is similar to, and often confused with, the llama. However, alpacas are often noticeably smaller than llamas.
The two animals are closely related and can successfully crossbreed.
Both species are believed to have been domesticated from their wild relatives, the vicuña and guanaco.
There are two breeds of alpaca: the Suri alpaca and the Huacaya alpaca.

Alpacas are kept in herds that graze on the level heights of the Andes of Southern Peru, Western Bolivia, Ecuador, and Northern Chile at an altitude of 3,500 to 5,000 metres (11,000 to 16,000 feet) above sea level.
Alpacas are considerably smaller than llamas, and unlike llamas, they were not bred to be working animals, but were bred specifically for their fiber.
Alpaca fiber is used for making knitted and woven items, similar to sheep's wool.
These items include blankets, sweaters, hats, gloves, scarves, a wide variety of textiles, and ponchos, in South America, as well as sweaters, socks, coats, and bedding in other parts of the world.
The fiber comes in more than 52 natural colors as classified in Peru, 12 as classified in Australia, and 16 as classified in the United States.

Alpacas communicate through body language.
The most common is spitting to show dominance when they are in distress, fearful, or feel agitated.
Male alpacas are more aggressive than females, and tend to establish dominance within their herd group.
In some cases, alpha males will immobilize the head and neck of a weaker or challenging male in order to show their strength and dominance.

In the textile industry, "alpaca" primarily refers to the hair of Peruvian alpacas, but more broadly it refers to a style of fabric originally made from alpaca hair, such as mohair, Icelandic sheep wool, or even high-quality wool from other breeds of sheep.
In trade, distinctions are made between alpacas and the several styles of mohair and luster.

An adult alpaca generally is between 81 and 99 centimetres (32 and 39 inches) in height at the shoulders (withers).
They usually weigh between 48 and 90 kilograms (106 and 198 pounds).
Raised in the same conditions, the difference in weight can be small with males weighting around 22.3 kilograms (49 lb 3 oz) and females 21.3 kilograms (46 lb 15 oz).
""".strip()
]

In [7]:
def model_perplexity(model, input_ids):
    input_ids_torch = torch.LongTensor([input_ids]).to(model.device)
    model.eval()
    with torch.no_grad():
        logits_sentence = model(input_ids_torch).logits[0]
        log_softmax_sentence = torch.nn.functional.log_softmax(logits_sentence, dim=-1)
    log_softmax_sentence_np = log_softmax_sentence.detach().cpu().numpy()
    next_tokens = input_ids[1:]
    log_softmax_correct_tokens = np.array([
        log_softmax_sentence_np[i, token]
        for i, token in enumerate(next_tokens)
    ])
    return np.exp(-log_softmax_correct_tokens.mean())

In [8]:
perplexities_original = [
    model_perplexity(model,
                     tokenizer(text)['input_ids'])
    for text in TEXTS
]
perplexities_original

[6.494379, 5.0919175, 4.490609]

In [9]:
lora_config = LoraConfig(
    r=8,
    lora_alpha=16,
    target_modules=["q_proj", "v_proj"],
    lora_dropout=0.05,
    bias="none",
    task_type="CAUSAL_LM",
)
lora_model = get_peft_model(model, lora_config)

In [10]:
for module in lora_model.modules():
    if 'Autograd4bitQuantLinear' in str(type(module)) or 'Linear4bitLt' in str(type(module)):
        if hasattr(module, "is_v1_model") and module.is_v1_model:
            module.zeros = module.zeros.half()
        module.scales = module.scales.half()

In [11]:
tokenizer.pad_token_id = 0

In [12]:
from alpaca_lora_4bit.gradient_checkpointing import apply_gradient_checkpointing
apply_gradient_checkpointing(lora_model, checkpoint_ratio=1.0);

Forward Patch Applied For Block 0
Forward Patch Applied For Block 1
Forward Patch Applied For Block 2
Forward Patch Applied For Block 3
Forward Patch Applied For Block 4
Forward Patch Applied For Block 5
Forward Patch Applied For Block 6
Forward Patch Applied For Block 7
Forward Patch Applied For Block 8
Forward Patch Applied For Block 9
Forward Patch Applied For Block 10
Forward Patch Applied For Block 11
Forward Patch Applied For Block 12
Forward Patch Applied For Block 13
Forward Patch Applied For Block 14
Forward Patch Applied For Block 15
Forward Patch Applied For Block 16
Forward Patch Applied For Block 17
Forward Patch Applied For Block 18
Forward Patch Applied For Block 19
Forward Patch Applied For Block 20
Forward Patch Applied For Block 21
Forward Patch Applied For Block 22
Forward Patch Applied For Block 23
Forward Patch Applied For Block 24
Forward Patch Applied For Block 25
Forward Patch Applied For Block 26
Forward Patch Applied For Block 27
Forward Patch Applied For Bloc

In [13]:
from torch.utils.data import Dataset


class ExperimentDataset:
    def __init__(self, encoded_input_ids) -> None:
        self.input_ids = [
            item[:-1]
            for item in encoded_input_ids
        ]
        self.labels = [
            item[1:]
            for item in encoded_input_ids
        ]
    
    def __len__(self):
        return len(self.input_ids)

    def __getitem__(self, index):
        return {
            "input_ids": torch.LongTensor(self.input_ids[index]),
            "labels": torch.LongTensor(self.labels[index]),
        }

In [14]:
dataset = ExperimentDataset(encoded_input_ids=[
    tokenizer(text)['input_ids']
    for text in TEXTS
])
len(dataset)

3

In [16]:
perplexities_nontuned_lora = [
    model_perplexity(lora_model,
                     tokenizer(text)['input_ids'])
    for text in TEXTS
]
perplexities_nontuned_lora

[6.494379, 5.0919175, 4.490609]

In [17]:
from transformers import TrainingArguments, Trainer

training_arguments = TrainingArguments(
    per_device_train_batch_size=1,
    per_device_eval_batch_size=1,
    gradient_accumulation_steps=1,
    warmup_steps=10,
    optim="adamw_torch",
    num_train_epochs=10,
    learning_rate=2e-4,
    fp16=True,
    logging_steps=1,
    evaluation_strategy="epoch",
    save_strategy="no",
    output_dir="test-finetune-loss-perplexity",
    report_to="none",
)

trainer = Trainer(
    model=model,
    train_dataset=dataset,
    eval_dataset=dataset,
    args=training_arguments,
)
model.config.use_cache = False

In [18]:
trainer.evaluate()

100%|██████████| 3/3 [00:02<00:00,  1.17it/s]


{'eval_loss': 13.802539825439453,
 'eval_runtime': 2.9367,
 'eval_samples_per_second': 1.022,
 'eval_steps_per_second': 1.022}

In [19]:
trainer.train()

{'loss': 14.1635, 'learning_rate': 2e-05, 'epoch': 0.33}


{'loss': 13.9738, 'learning_rate': 4e-05, 'epoch': 0.67}


{'loss': 13.2638, 'learning_rate': 6e-05, 'epoch': 1.0}


 67%|██████▋   | 2/3 [00:00<00:00,  2.83it/s]
                                              


{'eval_loss': 13.769198417663574, 'eval_runtime': 3.558, 'eval_samples_per_second': 0.843, 'eval_steps_per_second': 0.843, 'epoch': 1.0}


100%|██████████| 3/3 [00:03<00:00,  2.01it/s]


{'loss': 13.2391, 'learning_rate': 8e-05, 'epoch': 1.33}


{'loss': 14.081, 'learning_rate': 0.0001, 'epoch': 1.67}


{'loss': 13.8237, 'learning_rate': 0.00012, 'epoch': 2.0}


 67%|██████▋   | 2/3 [00:00<00:00,  2.58it/s]
                                              


{'eval_loss': 13.514114379882812, 'eval_runtime': 3.733, 'eval_samples_per_second': 0.804, 'eval_steps_per_second': 0.804, 'epoch': 2.0}


100%|██████████| 3/3 [00:03<00:00,  1.81it/s]


{'loss': 13.7035, 'learning_rate': 0.00014, 'epoch': 2.33}


{'loss': 12.8097, 'learning_rate': 0.00016, 'epoch': 2.67}


{'loss': 13.4416, 'learning_rate': 0.00018, 'epoch': 3.0}


 67%|██████▋   | 2/3 [00:00<00:00,  2.58it/s]
                                              


{'eval_loss': 12.753833770751953, 'eval_runtime': 3.713, 'eval_samples_per_second': 0.808, 'eval_steps_per_second': 0.808, 'epoch': 3.0}


100%|██████████| 3/3 [00:03<00:00,  1.82it/s]


{'loss': 12.2504, 'learning_rate': 0.0002, 'epoch': 3.33}


{'loss': 12.4893, 'learning_rate': 0.00019, 'epoch': 3.67}


{'loss': 12.0147, 'learning_rate': 0.00018, 'epoch': 4.0}


 67%|██████▋   | 2/3 [00:00<00:00,  2.60it/s]
                                               


{'eval_loss': 11.133675575256348, 'eval_runtime': 3.73, 'eval_samples_per_second': 0.804, 'eval_steps_per_second': 0.804, 'epoch': 4.0}


100%|██████████| 3/3 [00:03<00:00,  1.82it/s]


{'loss': 10.7345, 'learning_rate': 0.00017, 'epoch': 4.33}


{'loss': 10.7954, 'learning_rate': 0.00016, 'epoch': 4.67}


{'loss': 10.0947, 'learning_rate': 0.00015000000000000001, 'epoch': 5.0}


 67%|██████▋   | 2/3 [00:11<00:05,  5.59s/it]
                                               


{'eval_loss': 9.410979270935059, 'eval_runtime': 40.0241, 'eval_samples_per_second': 0.075, 'eval_steps_per_second': 0.075, 'epoch': 5.0}


100%|██████████| 3/3 [00:39<00:00,  7.71s/it]


{'loss': 9.654, 'learning_rate': 0.00014, 'epoch': 5.33}


{'loss': 8.8581, 'learning_rate': 0.00014, 'epoch': 5.67}


{'loss': 8.7478, 'learning_rate': 0.00013000000000000002, 'epoch': 6.0}


 67%|██████▋   | 2/3 [00:07<00:03,  3.81s/it]
                                               


{'eval_loss': 8.475296020507812, 'eval_runtime': 29.1612, 'eval_samples_per_second': 0.103, 'eval_steps_per_second': 0.103, 'epoch': 6.0}


100%|██████████| 3/3 [00:28<00:00,  5.46s/it]


{'loss': 8.4006, 'learning_rate': 0.00012, 'epoch': 6.33}


{'loss': 7.9889, 'learning_rate': 0.00011000000000000002, 'epoch': 6.67}


{'loss': 7.9014, 'learning_rate': 0.0001, 'epoch': 7.0}


 67%|██████▋   | 2/3 [00:07<00:03,  3.79s/it]
                                               


{'eval_loss': 7.3786940574646, 'eval_runtime': 29.1281, 'eval_samples_per_second': 0.103, 'eval_steps_per_second': 0.103, 'epoch': 7.0}


100%|██████████| 3/3 [00:28<00:00,  5.45s/it]


{'loss': 7.5275, 'learning_rate': 9e-05, 'epoch': 7.33}


{'loss': 6.9847, 'learning_rate': 8e-05, 'epoch': 7.67}


{'loss': 6.824, 'learning_rate': 7e-05, 'epoch': 8.0}


 67%|██████▋   | 2/3 [00:06<00:03,  3.32s/it]
                                               


{'eval_loss': 6.589598178863525, 'eval_runtime': 24.4992, 'eval_samples_per_second': 0.122, 'eval_steps_per_second': 0.122, 'epoch': 8.0}


100%|██████████| 3/3 [00:24<00:00,  4.78s/it]


{'loss': 6.555, 'learning_rate': 6e-05, 'epoch': 8.33}


{'loss': 6.4221, 'learning_rate': 5e-05, 'epoch': 8.67}


{'loss': 6.351, 'learning_rate': 4e-05, 'epoch': 9.0}


 67%|██████▋   | 2/3 [00:06<00:03,  3.39s/it]
                                               


{'eval_loss': 6.250532150268555, 'eval_runtime': 25.3818, 'eval_samples_per_second': 0.118, 'eval_steps_per_second': 0.118, 'epoch': 9.0}


100%|██████████| 3/3 [00:25<00:00,  4.84s/it]


{'loss': 6.298, 'learning_rate': 3e-05, 'epoch': 9.33}


{'loss': 6.2138, 'learning_rate': 2e-05, 'epoch': 9.67}


{'loss': 6.1457, 'learning_rate': 1e-05, 'epoch': 10.0}


 67%|██████▋   | 2/3 [00:06<00:03,  3.35s/it]
                                               


{'eval_loss': 6.17491340637207, 'eval_runtime': 25.4357, 'eval_samples_per_second': 0.118, 'eval_steps_per_second': 0.118, 'epoch': 10.0}


100%|██████████| 3/3 [00:25<00:00,  4.75s/it]


{'train_runtime': 674.6923, 'train_samples_per_second': 0.044, 'train_steps_per_second': 0.044, 'train_loss': 10.058381668726604, 'epoch': 10.0}


100%|██████████| 30/30 [11:14<00:00, 22.49s/it]


TrainOutput(global_step=30, training_loss=10.058381668726604, metrics={'train_runtime': 674.6923, 'train_samples_per_second': 0.044, 'train_steps_per_second': 0.044, 'train_loss': 10.058381668726604, 'epoch': 10.0})

In [20]:
trainer.evaluate()

100%|██████████| 3/3 [00:25<00:00,  8.58s/it]


{'eval_loss': 6.17491340637207,
 'eval_runtime': 26.0902,
 'eval_samples_per_second': 0.115,
 'eval_steps_per_second': 0.115,
 'epoch': 10.0}

In [21]:
perplexities_tuned_lora = [
    model_perplexity(lora_model,
                     tokenizer(text)['input_ids'])
    for text in TEXTS
]
perplexities_tuned_lora

[26.089653, 23.530462, 32.777637]

## Results

So:
- initial model perplexities on this texts were: 
    `[6.494379, 5.0919175, 4.490609]`
- perplexities after applying LoRA adapters, but before finetuning (basically the same model with some noise?)
    `[6.494379, 5.0919175, 4.490609]`
- loss of LoRA-applied model before finetuning:
    ```
    {'eval_loss': 13.802539825439453,
    'eval_runtime': 2.9367,
    'eval_samples_per_second': 1.022,
    'eval_steps_per_second': 1.022}
    ```
- loss of LoRA model after trying to overfit for this texts:
    ```
    {'eval_loss': 6.17491340637207,
    'eval_runtime': 26.0902,
    'eval_samples_per_second': 0.115,
    'eval_steps_per_second': 0.115,
    'epoch': 10.0}
    ```
- perplexities
    ```
    [26.089653, 23.530462, 32.777637]
    ```